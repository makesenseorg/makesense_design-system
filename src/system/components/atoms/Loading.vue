<template>
  <div :class="['loading', 'loading--' + type, { 'loading--inline': inline }]">
    <svg-loading class="loading__svg" :width="size" />
    <div class="loading__progress" v-if="progress !== -1">
      {{ getProgress() }}%
    </div>
  </div>
</template>

<script>
import svgLoading from "@@/assets/img/ui/loading.svg";
/**
 * Displays a loading indicator
 * @version 1.0.0
 */
export default {
  name: "MksLoading",
  components: {
    svgLoading
  },
  props: {
    /**
     * height of the zone covered by the loading indicator. Full by default
     *  "full": whole page, "component": 60vh, "standalone": auto, "simple": just the size of loading as a block element
     */
    type: {
      type: String,
      default: "full",
      validator: function(value) {
        return (
          ["full", "component", "standalone", "simple"].indexOf(value) !== -1
        );
      }
    },
    /**
     * Progress of the loading
     */
    progress: {
      type: Number,
      default: -1
    },
    /**
     * Size of the loading indicator in px
     */
    size: {
      type: Number,
      default: 50
    },
    /**
     * Is it inline ?
     */
    inline: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    getProgress: function() {
      return Math.floor(this.progress);
    }
  }
};
</script>

<style lang="scss" scoped>
.loading {
  position: relative;

  &--full {
    height: 100vh;
    width: 100%;
    background: $color-neutral-100;
  }

  &--component {
    height: 60vh;
  }

  &--standalone {
    height: auto;
  }

  &:not(.loading--simple) {
    @include flex;
    align-items: center;
    justify-content: center;
  }

  &--inline {
    display: inline-block;
    vertical-align: middle;
  }
}

.loading__progress {
  position: absolute;
  @include text-small;
}
</style>
<docs>
## Default loading - (full)

```jsx
<mks-loading/>
```

## Component loading

```jsx
<mks-loading type="component" />
```

## Standalone loading

```jsx
<mks-loading type="standalone" />
```

## Inline loading

```jsx
<p>Sending... <mks-loading v-bind:inline="true" type="simple" v-bind:size="30"/></p>
```
</docs>
