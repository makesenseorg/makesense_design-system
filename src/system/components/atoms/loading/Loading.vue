<template>
  <div :class="['loading', 'loading--' + type, { 'loading--inline': inline }]">
    <svg
      class="loading__svg"
      :width="size"
      :height="size"
      viewBox="-3 -3 44 44"
      xmlns="http://www.w3.org/2000/svg"
      :stroke="color"
    >
      <g fill="none" fill-rule="evenodd">
        <g transform="translate(1 1)" :stroke-width="size < 25 ? 5 : 3">
          <circle stroke-opacity="0.5" cx="18" cy="18" r="18" :stroke="color" />
          <path d="M36 18c0-9.94-8.06-18-18-18">
            <animateTransform
              attributeName="transform"
              type="rotate"
              from="0 18 18"
              to="360 18 18"
              dur="1s"
              repeatCount="indefinite"
            />
          </path>
        </g>
      </g>
    </svg>
    <div class="loading__progress" v-if="progress !== -1">
      {{ getProgress() }}%
    </div>
  </div>
</template>

<script>
import svgLoading from "@@/assets/img/ui/loading.svg";
/**
 * Displays a loading indicator
 * @version 1.0.0
 */
export default {
  name: "MksLoading",
  components: {
    svgLoading,
  },
  props: {
    /**
     * height of the zone covered by the loading indicator. Full by default
     *  "full": whole page, "component": 60vh, "standalone": auto, "simple": just the size of loading as a block element
     */
    type: {
      type: String,
      default: "full",
      validator: function(value) {
        return (
          ["full", "component", "standalone", "simple"].indexOf(value) !== -1
        );
      },
    },
    /**
     * Progress of the loading
     */
    progress: {
      type: Number,
      default: -1,
    },
    /**
     * Size of the loading indicator in px
     */
    size: {
      type: Number,
      default: 50,
    },
    /**
     * Is it inline ?
     */
    inline: {
      type: Boolean,
      default: false,
    },
    /**
     * Color
     */
    color: {
      type: String,
      default: "#3c4858",
    },
  },
  methods: {
    getProgress: function() {
      return Math.floor(this.progress);
    },
  },
};
</script>

<style lang="scss" scoped>
.loading {
  position: relative;

  &--full {
    height: 100vh;
    width: 100%;
    background: $color-neutral-100;
  }

  &--component {
    height: 60vh;
  }

  &--standalone {
    height: auto;
  }

  &:not(.loading--simple) {
    @include flex;
    align-items: center;
    justify-content: center;
  }

  &--inline {
    display: inline-block;
    vertical-align: middle;
  }
}

.loading__progress {
  position: absolute;
  @include text-small;
}
</style>
<docs>
## Default loading - (full)

```jsx
<mks-loading/>
```

## Component loading

```jsx
<mks-loading type="component" color="#ff0000" />
```

## Standalone loading

```jsx
<mks-loading type="standalone" />
```

## Inline loading

```jsx
<p>Sending... <mks-loading v-bind:inline="true" type="simple" v-bind:size="30"/></p>
```
</docs>
