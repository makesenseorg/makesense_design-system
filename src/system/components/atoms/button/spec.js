import { shallowMount, createLocalVue } from "@vue/test-utils";
import Comp from "./Button.vue";
import Loading from "../loading/Loading.vue";

const localVue = createLocalVue();
localVue.component("MksLoading", Loading);


const mutationObserverMock = jest.fn(function MutationObserver(callback) {
    this.observe = jest.fn();
    this.disconnect = jest.fn();
    // Optionally add a trigger() method to manually trigger a change
    this.trigger = (mockedMutationsList) => {
        callback(mockedMutationsList, this);
    };
});
global.MutationObserver = mutationObserverMock;

describe("Button.vue", () => {
  it("matches snapshot", () => {
    const wrapper = shallowMount(Comp);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("emits click event", async () => {
    const wrapper = shallowMount(Comp);

    wrapper.trigger("click");
    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().click).toBeTruthy();
  });

  it("sets wrapper tag", () => {
    const wrapper = shallowMount(Comp, {
      propsData: {
        tag: "a",
      },
    });
    expect(wrapper.element.tagName).toBe("A");
  });

  it("has loading style", () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        loading: true,
      },
    });
    const loading = wrapper.findComponent(Loading);
    expect(loading.exists()).toBe(true);
    expect(wrapper.classes()).toContain("button--loading");
  });

  // test type
  it("type primary by default", () => {
    const wrapper = shallowMount(Comp);
    expect(wrapper.classes()).toContain("button--type-primary");
  });

  it("sets class with type modifier", () => {
    const wrapper = shallowMount(Comp, {
      propsData: {
        type: "secondary",
      },
    });
    expect(wrapper.classes()).toContain("button--type-secondary");
  });

  // test size
  it("sets size default by default", () => {
    const wrapper = shallowMount(Comp);
    expect(wrapper.classes()).toContain("button--size-default");
  });

  it("sets class with size modifier", () => {
    const wrapper = shallowMount(Comp, {
      propsData: {
        size: "small",
      },
    });
    expect(wrapper.classes()).toContain("button--size-small");
  });

  // test inputType
  it("sets input type attribute", () => {
    const wrapper = shallowMount(Comp, {
      propsData: {
        inputType: "submit",
      },
    });
    expect(wrapper.attributes('type')).toBe("submit");
  });

  // test slot
  it("displays slot content", () => {
    const wrapper = shallowMount(Comp, {
      slots: {
        default: "clickme",
      },
    });

    expect(wrapper.text()).toBe("clickme");
  });


  // test minwidth

  
});