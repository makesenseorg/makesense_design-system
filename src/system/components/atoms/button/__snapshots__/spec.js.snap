// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue matches snapshot 1`] = `
<div
  class="button button--type-primary button--size-default"
  role="button"
  style="min-width: 0px;"
  tabindex="0"
  type=""
/>
`;
